src/mame/misc/orientp.cpp

```diff
@@ -0,0 +1,280 @@
+// license:BSD-3-Clause
+// copyright-holders:flama12333
+/*************************************************************************
+// 
+
+
+ notes: 
+The dump of set 2 was from Soccer Santiago II 6 ball pinball I dont known which was from due to unconfirmed, so im gonna to name as unknown.
+
+these roulettes use a 2x8 scan matrix for keyboard with the 8279. Only 14 buttons are used.
+
+Port at back pcb:
+small port: bonus PCB (7 segment 4-digit)
+big port: playfield sensors PCB with lots of 74HC245, leds, sensors, multiplex
+
+Hardware info - may not accurate
+Buttons
+K1
+K2
+K3
+K4
+east8
+ic
+u1 kc8279
+u17 and u21 nec d8255ac-2
+u32 hm6264
+??  winbond w27c020? adpcm rom.
+u33 winbond w27c512 boot rom.
+u39 at89s51 second mcu for protection.
+
+// TODO:
+Need proper description.
+Error code 10
+Hardware info.
+MCU Hook up
+Verify memory maps.
+Need Proper Layout .
+
+*/
+
+#include "emu.h"
+
+#include "cpu/mcs51/mcs51.h"
+#include "machine/nvram.h"
+#include "machine/i8255.h"
+#include "machine/i8279.h"
+#include "sound/ay8910.h"
+#include "sound/okim6295.h"
+#include "sound/ymopl.h"
+#include "orientp.lh"
+#include "speaker.h"
+
+namespace {
+
+class orientalpearl_state : public driver_device
+{
+public:
+	orientalpearl_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag)
+		, m_digits(*this, "digit%u", 0U)
+		, m_leds(*this, "led%u", 0U)
+		, m_inputs(*this, { "KEYS1", "KEYS2", "DSW1", "PUSHBUTTONS" })
+	{ }
+
+	void orientp(machine_config &config);
+
+private:
+	void io_map(address_map &map);
+	void program_map(address_map &map);
+	void mcu_io_map(address_map &map);
+	void mcu_map(address_map &map);
+    void display_7seg_data_w(uint8_t data);
+	void multiplex_7seg_w(uint8_t data);
+	void ay1_port_a_w(uint8_t data);
+	void ay1_port_b_w(uint8_t data);
+	uint8_t keyboard_r();
+	uint8_t m_selected_7seg_module = 0;
+
+	output_finder<32> m_digits;
+	output_finder<30> m_leds;
+	required_ioport_array<4> m_inputs;
+
+protected:
+	virtual void machine_start() override;
```

Comment:
Cold lifecyle member functions should be marked with `ATTR_COLD`.

The `protected` section should come before the `private` section if there isn’t a compelling reason not to.